src/west/manifest.py

```diff
@@ -111,17 +337,68 @@ def validate(data):
         min_version_str = str(data['version'])
         min_version = parse_version(min_version_str)
         if min_version > _SCHEMA_VER:
-            raise ManifestVersionError(min_version)
+            raise ManifestVersionError(min_version_str)
         elif min_version < _EARLIEST_VER:
             raise MalformedManifest(
-                f'invalid version {min_version}; '
+                f'invalid version {min_version_str}; '
                 f'lowest schema version is {_EARLIEST_VER_STR}')
 
     try:
         pykwalify.core.Core(source_data=data,
                             schema_files=[_SCHEMA_PATH]).validate()
     except pykwalify.errors.SchemaError as se:
-        raise MalformedManifest(se._msg) from se
+        raise MalformedManifest(se.msg) from se
+
+#
+# Exception types
+#
+
+class MalformedManifest(Exception):
+    '''Manifest parsing failed due to invalid data.
+    '''
+
+class MalformedConfig(Exception):
+    '''The west configuration was malformed in a way that made a
+    manifest operation fail.
+    '''
+
+class ManifestImportFailed(Exception):
+    '''An operation required to resolve a manifest failed.
+
+    Attributes:
+
+    - ``project``: the Project instance with the missing manifest data
+    - ``filename``: the missing file
+    '''
+
+    def __init__(self, project: 'Project', filename: str):
+        self.project = project
+        self.filename = filename
```

Comment:
Better use `@property`

Reply:
Could you provide some rationale why that would be better? These are just simple attributes.

Reply:
`properties` have explicit get (r) or get/set (rw). I tend to encapsulate class data as properties, it's ok as it is though :-)